src/utils/ratatui.rs

```diff
@@ -0,0 +1,59 @@
+use std::{
+    io::{Result, Write},
+    sync::{Arc, Mutex},
+};
+
+use ratatui::{buffer::Buffer, layout::Rect, widgets::Widget};
+
+use crate::Stream;
+
+/// A ratatui widget that can be used to display a spinner
+///
+/// # Examples
+///
+/// ```no_run
+/// use ratatui::widgets::SpinnerWidget;
+///
+/// let spinner_widget = SpinnerWidget::default();
+/// let spinner = Spinner::with_stream(
+///     Spinners::Dots9,
+///     "Waiting ".into(),
+///     spinner_widget.as_steam(),
+/// );
+/// # let area = Rect::default();
+/// # let buf = &mut Buffer::empty(area);
+/// spinner_widget.render(area, buf);
+/// ```
+#[derive(Clone, Default)]
+pub struct SpinnerWidget {
+    value: Arc<Mutex<Vec<u8>>>,
```

Comment:
It's possible that this could have been the shared state instead of the entire widget, but I prefer passing around wrapped shared state values instead of raw values. Do you have preferences one way or another?